### ISSUE TITLE
S3 checksum mismatch when backing up database

### ISSUE THREAD

**Issue by ZoidBB** [pgbackrest/pgbackrest#738]
Please provide the following information when submitting an issue (feature requests or general comments can skip this):

1. pgBackRest version:
2.10

2. PostgreSQL version:
9.5.16

3. Operating system/version - if you have more than one server (for example, a database server, a repository host server, one or more standbys), please specify each:
Amazon Linux 1

4. Did you install pgBackRest from source or from a package?
RPM package

5. Please attach the following as applicable:
[global]
process-max=32
repo1-cipher-type=none
repo1-path=/prod
repo1-retention-diff=2
repo1-retention-full=2
repo1-s3-endpoint=s3.amazonaws.com
repo1-s3-bucket=REDACTED
repo1-s3-region=REDACTED
repo1-type=s3
repo1-s3-key=REDACTED
repo1-s3-key-secret=REDACTED
repo1-s3-ca-file=/etc/pki/tls/certs/ca-bundle.crt
start-fast=y
stop-auto=y

[global:archive-push]
compress-level=3

[main]
pg1-path=/pgdata/data

[backup]
pg1-path=/pgdata_snapshot/data

7. Describe the issue:
Intermittent error when running a backup of our ~7TB database, indicating mismatched checksums when uploading a chunk to S3

```
ERROR [039]: : raised from local-18 process: S3 request error [400] Bad
*** request header ***
PUT /prod/backup/backup/20190604-112514F_20190605-104038I/pg_data/base/16387/144507543.63.gz?partNumber=10&uploadId=0J8PftvIMxAAYsERz7Qb.0lVMPx3NIHQgjXNz5iODVDWGgritYEmYvq7TA5JO9rsBR8AKM36h2tPoL69X0MoC.b6figgPEurUYF1ipug_dT5zzo0eD8ufcF9nMd1kfP3 HTTP/1.1
authorization: <redacted>
content-length: 17713575
content-md5: tpF9hUR43M/hqAWfQM0g5Q==
host: ccodb-backup-763429161784-prod-us-west-2.s3.amazonaws.com
x-amz-content-sha256: 76e1a4ccff658f6b80bc12b11a7987406ede2d55e82d9f81103d3d43c9aff744
x-amz-date: 20190605T143351Z
*** response header ***
x-amz-request-id: E668FE42B273B04D
x-amz-id-2: zBwOcbHd5rcrv4HmYOf06lpPX8T93t8+Als/HGxlqY7AdpT6tz009P7yNCHKab++EETiiWGj11E=
Content-Type: application/xml
Transfer-Encoding: chunked
Date: Wed, 05 Jun 2019 14:33:55 GMT
Connection: close
Server: AmazonS3
*** response body ***
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>XAmzContentSHA256Mismatch</Code><Message>The provided 'x-amz-content-sha256' header does not match what was computed.</Message><ClientComputedContentSHA256>76e1a4ccff658f6b80bc12b11a7987406ede2d55e82d9f81103d3d43c9aff744</ClientComputedContentSHA256><S3ComputedContentSHA256>3d865108147cf2d01a789007b54497de9d6140c3e115ebc600437d0e74e9bd6a</S3ComputedContentSHA256><RequestId>E668FE42B273B04D</RequestId><HostId>zBwOcbHd5rcrv4HmYOf06lpPX8T93t8+Als/HGxlqY7AdpT6tz009P7yNCHKab++EETiiWGj11E=</HostId></Error>
```

**Comment by dwsteele** [comment]
It looks like the XML in the response body got mangled so it's hard to tell what's going on.  Can you paste that inside a code block to make it clearer?

But the first N characters of the message seem to be the exact sha256 hash we generated, so that's odd.  I wonder if this is something that needs to be retried.

**Comment by ZoidBB** [comment]
Whoops! Description updated.

The interesting bits are:
`<ClientComputedContentSHA256>76e1a4ccff658f6b80bc12b11a7987406ede2d55e82d9f81103d3d43c9aff744</ClientComputedContentSHA256>`
`<S3ComputedContentSHA256>3d865108147cf2d01a789007b54497de9d6140c3e115ebc600437d0e74e9bd6a</S3ComputedContentSHA256>`

The hash being sent by pgbackrest doesn't match what S3 is computing.

**Comment by dwsteele** [comment]
Ah yes, now I see it.  Does this happen in the same place on a retry or is it more or less random?

**Comment by grayaii** [comment]
@dwsteele It is random. 
This seems to happen whether or not compression is enabled during a backup, but it seems to be happening more often than not when compression is enabled. 
Our database is almost 7 terabytes large (it's HUGE).  Sometimes the backup works, sometimes it doesn't.

I'm running stress test right now on 3 different instances (one with compression enabled, one with compression disabled, and one with pgbackerst 2.14 and compression enabled).

They all fail sporadically, but when compression is disabled, it fails less often.
Hope this helps!

**Comment by dwsteele** [comment]
The next release of pgBackRest is planned to use the C storage layer for all operations.  Right now backup is still using the Perl storage layer.

We'll want to see if it reproduces in the new code before moving forward on a fix.

**Comment by ZoidBB** [comment]
@dwsteele is there a test branch we can install and run some tests with?

**Comment by dwsteele** [comment]
Not yet -- I'll let you know when we have something stable.

**Comment by grayaii** [comment]
We're dying to get our hands on it to do some internal testing :)
@dwsteele If you have a branch we can play with, we can give you quick results.

**Comment by dwsteele** [comment]
Here's the PR: https://github.com/pgbackrest/pgbackrest/pull/750

We are still doing review and it needs a rebase but all the tests are passing.  I'll do the rebase in the next day or two once I have addressed review comments but you are welcome to try it out.

**Comment by dwsteele** [comment]
This branch has now been committed to master at 4815752ccc46ff742f67b369.

**Comment by dwsteele** [comment]
Any luck on testing this?

**Comment by grayaii** [comment]
Not yet, but it's at the top of my list to this!

**Comment by dwsteele** [comment]
Closing since there have been no updates.  The new storage code has been released since 2.16 so let us know if that is not working.
